tests/webhooks_receiver.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 Sorint.lab
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"testing"
+
+	"agola.io/agola/internal/errors"
+	"agola.io/agola/internal/services/notification"
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/internal/util"
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+
+	rstypes "agola.io/agola/services/runservice/types"
+)
+
+type webhooksReceiver struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	log zerolog.Logger
+
+	t   *testing.T
+	dir string
+
+	mu sync.Mutex
+
+	listenAddress string
+	exposedURL    string
+}
+
+func setupWebhooksReceiver(pctx context.Context, t *testing.T, dir string) *webhooksReceiver {
+	log := testutil.NewLogger(t)
+	ctx, cancel := context.WithCancel(pctx)
+
+	wr := &webhooksReceiver{ctx: ctx, t: t, dir: dir, log: log, cancel: cancel}
+
+	_, port, err := testutil.GetFreePort(true, false)
+	if err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	listenAddress, _, err := testutil.GetFreePort(true, false)
+	if err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	wr.listenAddress = fmt.Sprintf("%s:%s", listenAddress, port)
+	wr.exposedURL = fmt.Sprintf("http://%s:%s", listenAddress, port)
+
+	if err := wr.start(); err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	go func() {
+		<-ctx.Done()
+
+		wr.stop()
+	}()
+
+	return wr
+}
+
+func (wr *webhooksReceiver) start() error {
+	wr.mu.Lock()
+	defer wr.mu.Unlock()
+
+	runPhases := make([]rstypes.RunPhase, 0)
+
+	router := mux.NewRouter()
+
+	router.Handle("/runevents", NewCreateRunEventHandler(wr.log, &runPhases)).Methods("POST")
+	router.Handle("/runevents", NewRunEventsHandler(wr.log, &runPhases)).Methods("GET")
```

Comment:
I bet this will cause a race if run under the race detector.

Instead of passing a pointer to a slice to multiple handlers you should put the status in a dedicated struct and provide some methods to add and get received webhooks.